include/pros/ext_adi.h

```diff
@@ -0,0 +1,641 @@
+/**
+ * \file pros/ext_adi.h
+ *
+ * Contains prototypes for interfacing with the 3-Wire Expander.
+ *
+ * Visit https://pros.cs.purdue.edu/v5/tutorials/topical/adi.html to learn more.
+ *
+ * This file should not be modified by users, since it gets replaced whenever
+ * a kernel upgrade occurs.
+ *
+ * Copyright (c) 2017-2020, Purdue University ACM SIGBots.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#ifndef _PROS_EXT_ADI_H_
+#define _PROS_EXT_ADI_H_
+
+#include <stdbool.h>
+#include <stdint.h>
+#include "pros/adi.h"
+#include "adi.h"
+#ifndef PROS_ERR
+#define PROS_ERR (INT32_MAX)
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+namespace pros {
+#endif
+
+#ifdef __cplusplus
+namespace c {
+#endif
+
+/******************************************************************************/
+/**                         General ADI Use Functions                        **/
+/**                                                                          **/
+/**       These functions allow for interaction with any ADI port type       **/
+/******************************************************************************/
+
+/**
+ * Gets the configuration for the given ADI port.
+ *
+ * This function uses the following values of errno when an error state is
+ * reached:
+ * ENXIO - The given value is not within the range of ADI Ports.
+ *
+ * \param adi_port
+ *	      The ADI port number (from 1-8, 'a'-'h', 'A'-'H') for which to return
+ *        the configuration
+ * \param smart_port
+ *        The smart port number that the ADI Expander is in
+ *
+ * \return The ADI configuration for the given port
+ */
+adi_port_config_e_t ext_adi_port_get_config(uint8_t adi_port, uint8_t smart_port);
```

Comment:
Should `adi_port` and `smart_port` be switched for all the external adi functions to stay consistent with all the other smart port devices?

Reply:
since we're no longer using the default parameters method to distinguish between internal and external ADI objects in the C++ API I think it would be smart to go back to `smart_port, adi_port` as the idiom (of course this would need to change in the port pair and port tuple "types" as well)

Reply:
I agree with this, as it makes more sense from a logical standpoint as well - when defining the path from the brain to the port, you start with the smart port and end with the ADI port.